src/Processors/QueryPlan/JoinStepLogical.cpp

```diff
@@ -781,4 +785,100 @@ std::optional<ActionsDAG> JoinStepLogical::getFilterActions(JoinTableSide side,
     return {};
 }
 
+void JoinStepLogical::serializeSettings(QueryPlanSerializationSettings & settings) const
+{
+    join_settings.updatePlanSettings(settings);
+    sorting_settings.updatePlanSettings(settings);
+}
+
+void JoinStepLogical::serialize(Serialization & ctx) const
+{
+    if (prepared_join_storage)
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Serialization of JoinStepLogical with prepared storage is not implemented");
+
+    if (swap_inputs)
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Serialization of JoinStepLogical with swapped inputs is not implemented");
+
+    UInt8 flags = 0;
+    if (use_nulls)
+        flags |= 1;
+
+    writeIntBinary(flags, ctx.out);
+
+    JoinActionRef::ActionsDAGRawPtrs dags = {
+        expression_actions.left_pre_join_actions.get(),
+        expression_actions.right_pre_join_actions.get(),
+        expression_actions.post_join_actions.get()};
+
+    writeVarUInt(dags.size(), ctx.out);
+    for (const auto * dag : dags)
+        dag->serialize(ctx.out, ctx.registry);
+
+    join_info.serialize(ctx.out, dags);
+
+    writeVarUInt(required_output_columns.size(), ctx.out);
+    for (const auto & name : required_output_columns)
+        writeStringBinary(name, ctx.out);
+}
+
+std::unique_ptr<IQueryPlanStep> JoinStepLogical::deserialize(Deserialization & ctx)
+{
+    if (ctx.input_headers.size() != 2)
+        throw Exception(ErrorCodes::INCORRECT_DATA, "JoinStepLogical must have two input streams");
+
+    UInt8 flags;
+    readIntBinary(flags, ctx.in);
+
+    bool use_nulls = flags & 1;
```

Comment:
Can't we use value of `join_use_nulls` from corresponding settings?

Reply:
I think this flag is important enough to serialize it independently (cause it potentially changes the result type).
But I don't fully understand if we need it be cause, possibly, DAG may already have all the required type conversions.

The same question about required_output_columns (can we somehow calculate it from post DAG?)

Reply:
Currently `toNullable` added to DAGs when converting to physical, so logical step should be aware about the flag. The same with removing columns outside `required_output_columns`